drivers/kscan/kscan_gpio.c

```diff
@@ -0,0 +1,320 @@
+	}
+
+	for (int r = 0; r < MATRIX_ROWS; r++) {
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			bool pressed = read_state[r][c];
+
+			if (pressed != data->matrix_state[r][c]) {
+				data->matrix_state[r][c] = pressed;
+				data->callback(dev, r, c, pressed);
+			}
+		}
+	}
+
+	return 0;
+}
+
+#endif
+
+static void kscan_gpio_timer_handler(struct k_timer *timer)
+{
+	struct kscan_gpio_data *data =
+		CONTAINER_OF(timer, struct kscan_gpio_data, timer);
+
+	k_work_submit(&data->work);
+}
+
+static void kscan_gpio_work_handler(struct k_work *work)
+{
+	struct kscan_gpio_data *data =
+		CONTAINER_OF(work, struct kscan_gpio_data, work);
+
+	kscan_gpio_read(data->dev);
+}
+
+#define INIT_ROW_SCAN_ITEM(idx) \
+	data->rows[idx].device = device_get_binding(DT_GPIO_LABEL_BY_IDX(MATRIX_NODE_ID, row_gpios, idx)); \
+	if (data->rows[idx].device == NULL) { \
+		LOG_ERR("Unable to find row GPIO device"); \
+		return -EINVAL; \
+	} \
+	data->rows[idx].pin = DT_GPIO_PIN_BY_IDX(MATRIX_NODE_ID, row_gpios, idx); \
+	data->rows[idx].flags = DT_GPIO_FLAGS_BY_IDX(MATRIX_NODE_ID, row_gpios, idx); \
+	if (gpio_pin_configure(data->rows[idx].device, data->rows[idx].pin, row_gpio_dir | data->rows[idx].flags) < 0) { \
+		LOG_ERR("Unable to configure row GPIO pin"); \
+		return -EINVAL; \
+	}
+
+#define INIT_COL_SCAN_ITEM(idx) \
+	data->cols[idx].device = device_get_binding(DT_GPIO_LABEL_BY_IDX(MATRIX_NODE_ID, col_gpios, idx)); \
+	if (data->cols[idx].device == NULL) { \
+		LOG_ERR("Unable to find column GPIO device\n"); \
+		return -EINVAL; \
+	} \
+	data->cols[idx].pin = DT_GPIO_PIN_BY_IDX(MATRIX_NODE_ID, col_gpios, idx); \
+	data->cols[idx].flags = DT_GPIO_FLAGS_BY_IDX(MATRIX_NODE_ID, col_gpios, idx); \
+	if (gpio_pin_configure(data->cols[idx].device, data->cols[idx].pin, col_gpio_dir | data->cols[idx].flags) < 0) { \
+		LOG_ERR("Unable to configure col GPIO pin"); \
+		return -EINVAL; \
+	}
+
+static int kscan_gpio_configure(struct device *dev, kscan_callback_t callback)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+
+	if (!callback) {
+		return -EINVAL;
+	}
+
+	data->callback = callback;
+
+	return 0;
+}
+
+static int kscan_gpio_enable_callback(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+
+	k_timer_start(&data->timer,
+		      K_MSEC(CONFIG_KSCAN_GPIO_POLL_PERIOD),
+		      K_MSEC(CONFIG_KSCAN_GPIO_POLL_PERIOD));
+
+	return 0;
+}
+
+static int kscan_gpio_disable_callback(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+
+	k_timer_stop(&data->timer);
+
+	return 0;
+}
+
+
+static int kscan_gpio_init(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+
+#if DT_ENUM_IDX(MATRIX_NODE_ID, diode_direction) == 0
+	gpio_flags_t col_gpio_dir = GPIO_INPUT;
+	gpio_flags_t row_gpio_dir = GPIO_OUTPUT_INACTIVE;
+#else
+	gpio_flags_t col_gpio_dir = GPIO_OUTPUT_INACTIVE;
+	gpio_flags_t row_gpio_dir = GPIO_INPUT;
+#endif
+
+#if DT_PROP_HAS_IDX(MATRIX_NODE_ID, col_gpios, 0)
```

Comment:
This is the one piece I am most disliking about this implementation. Because the number of entries in the `col-gpios` and `row-gpios` is dynamic, we need to loop over the to initialize/load things, but there's no easy *macro* to do so.

#24536 is an open issue about this, and having a `DT_GPIO_FOREACH` would let me write this safely with something like:

```
DT_GPIOS_FOREACH(MATRIX_NODE_ID, col_gpios, INIT_COL_SCAN_ITEM)
```

and have it "Just Work"